lightning-invoice/src/utils.rs

```diff
@@ -5,22 +5,119 @@ use payment::{Payer, Router};
 use lightning::routing::network_graph::{NetworkGraph, RoutingFees};
 use lightning::routing::router::{Route, RouteHint, RouteHintHop, RouteParameters, find_route};
 use lightning::util::logger::Logger;
+use secp256k1::Secp256k1;
 use secp256k1::key::PublicKey;
 use core::convert::TryInto;
 use core::ops::Deref;
 use core::time::Duration;
 
+#[cfg(feature = "std")]
+/// Utility to create an invoice that can be paid to one of multiple nodes, or a "phantom invoice."
+/// This works because the invoice officially pays to a fake node (the "phantom"). Useful for
+/// load-balancing payments between multiple nodes.
+///
+/// `channels` contains a list of tuples of `(phantom_node_scid, real_node_pubkey,
+/// real_channel_details)`, where `phantom_node_scid` is unique to each `channels` entry and should
+/// be retrieved freshly for each new invoice from [`ChannelManager::get_phantom_scid`].
+/// `real_channel_details` comes from [`ChannelManager::list_channels`].
+///
+/// `payment_hash` and `payment_secret` come from [`ChannelManager::create_inbound_payment`] or
+/// [`ChannelManager::create_inbound_payment_for_hash`].
+///
+/// See [`KeysInterface::get_phantom_secret`] for more requirements on supporting phantom node
+/// payments.
+///
+/// [`KeysInterface::get_phantom_secret`]: lightning::chain::keysinterface::KeysInterface::get_phantom_secret
+pub fn create_phantom_invoice<Signer: Sign, K: Deref>(
+	amt_msat: Option<u64>, description: String, payment_hash: PaymentHash, payment_secret:
+	PaymentSecret, channels: &[(u64, PublicKey, &ChannelDetails)], keys_manager: K, network: Currency
+) -> Result<Invoice, SignOrCreationError<()>> where K::Target: KeysInterface {
+	if channels.len() == 0 {
+		return Err(SignOrCreationError::CreationError(CreationError::MissingRouteHints))
+	}
+	let phantom_secret = match keys_manager.get_phantom_secret(channels[0].0) {
+		Ok(s) => s,
+		Err(()) => return Err(SignOrCreationError::CreationError(CreationError::InvalidPhantomScid))
+	};
+	let mut route_hints = vec![];
+	for (phantom_scid, real_node_pubkey, channel) in channels {
+		let short_channel_id = match channel.short_channel_id {
+			Some(id) => id,
+			None => continue,
+		};
+		let forwarding_info = match &channel.counterparty.forwarding_info {
+			Some(info) => info.clone(),
+			None => continue,
+		};
+		route_hints.push(RouteHint(vec![
+			RouteHintHop {
+				src_node_id: channel.counterparty.node_id,
+				short_channel_id,
+				fees: RoutingFees {
+					base_msat: forwarding_info.fee_base_msat,
+					proportional_millionths: forwarding_info.fee_proportional_millionths,
+				},
+				cltv_expiry_delta: forwarding_info.cltv_expiry_delta,
+				htlc_minimum_msat: None,
+				htlc_maximum_msat: None,
+			},
+			RouteHintHop {
+				src_node_id: *real_node_pubkey,
+				short_channel_id: *phantom_scid,
+				fees: RoutingFees {
+					base_msat: 0,
+					proportional_millionths: 0,
+				},
+				cltv_expiry_delta: MIN_CLTV_EXPIRY_DELTA,
```

Comment:
If I understand this right, the phantom hop will increase the total time lock across the whole route. Is that necessary?

Reply:
True. One priority is that we don't want to make it obvious at-a-glance that an invoice is for a phantom node, so this does help with that I think

Reply:
Makes sense